subsys/emds/emds_new.c

```diff
@@ -254,16 +253,78 @@ int emds_load(void)
 	}
 
 	LOG_DBG("Found freshest snapshot in partition %d with fresh_cnt %u",
-		freshest_partition_idx, freshest_snapshot.fresh_cnt);
+		freshest_partition_idx, freshest_snapshot.metadata.fresh_cnt);
 
-	return emds_read_data(partition[freshest_partition_idx].fa, &freshest_snapshot);
+	return emds_read_data(partition[freshest_partition_idx].fa, &freshest_snapshot.metadata);
 }
 
 int emds_prepare(void)
 {
-	emds_ready = true;
+	size_t data_size;
+	bool erase_enabled = false;
+	int idx;
+	int freshest_partition_idx = -1;
+	int rc = 0;
 
-	return 0;
+	if (!emds_initialized) {
+		return -ECANCELED;
+	}
+
+	if (emds_ready) {
+		LOG_DBG("Already prepared");
+		return -EALREADY;
+	}
+
+	/* Returned status is not checked since initialization state is checked above */
+	(void)emds_store_size_get(&data_size);
+
+	/** First try to allocate snapshot in the same partition where freshest snapshot exists */
+	if (freshest_snapshot.metadata.fresh_cnt > 0) {
+		for (idx = 0; idx < PARTITIONS_NUM_MAX; idx++) {
+			if (IN_RANGE(freshest_snapshot.metadata_addr, partition[idx].fa->fa_off,
+				     partition[idx].fa->fa_off + partition[idx].fa->fa_size)) {
+				freshest_partition_idx = idx;
+				break;
+			}
+		}
+
+		rc = emds_flash_allocate_snapshot(&partition[idx], &freshest_snapshot,
+						  &allocated_snapshot, data_size);
+		if (rc == 0) {
+			emds_ready = true;
+			return 0;
+		}
+	}
+
```

Comment:
Also, can you not assume that if it is not possible to allocate new snapshot in above loop, then:

`idx = (freshest_partition_idx + 1) % PARTITIONS_NUM_MAX;`

and execute erase on this `idx` anyway? So partitions will keep getting consumed in circular order (once they fill up). That way all the complication below can be eliminated.

Reply:
I didn't get this.